src/models/helpers/genericCollaborator.js

```diff
@@ -0,0 +1,482 @@
+ *
+ * @param {string} genericCollaboratorType - entity type for collaborator.
+ * @param {Object} sequelize - The Sequelize instance.
+ * @param {Object} transaction - The transaction object.
+ * @param {string} typeName - The name of the collaborator type.
+ * @returns {Promise<Object>} - A promise that resolves to the found collaborator type ID.
+ */
+const getIdForCollaboratorType = async (
+  genericCollaboratorType,
+  sequelize,
+  transaction,
+  typeName,
+  // Find the collaborator type in the database
+) => sequelize.models.CollaboratorType.findOne({
+  where: {
+    name: typeName,
+  },
+  include: [{
+    model: sequelize.models.ValidFor,
+    as: 'validFor',
+    required: true,
+    attributes: [],
+    where: { name: collaboratorDetails[genericCollaboratorType].validFor },
+  }],
+  raw: true,
+  ...(transaction && { transaction }),
+});
+
+/**
+ * Creates a new entity collaborator in the database.
+ *
+ * @param {string} genericCollaboratorType - entity type for collaborator.
+ * @param {Object} sequelize - The Sequelize instance.
+ * @param {Object} transaction - The transaction object for atomicity.
+ * @param {number} entityId - The ID of the entity.
+ * @param {number} userId - The ID of the user.
+ * @param {string} typeName - The name of the collaborator type.
+ * @param {Object} linkBack - The link back object.
+ * @returns {Promise<Object>} - A Promise that resolves to the created entity collaborator.
+ */
+const createCollaborator = async (
+  genericCollaboratorType,
+  sequelize,
+  transaction,
+  entityId,
+  userId,
+  typeName,
+  linkBack = null,
+) => {
+  const { id: collaboratorTypeId } = await getIdForCollaboratorType(
+    genericCollaboratorType,
+    sequelize,
+    transaction,
+    typeName,
+  );
+  return sequelize.models[collaboratorDetails[genericCollaboratorType].collaborators]
+    .create({
+      [collaboratorDetails[genericCollaboratorType].idName]: entityId,
+      userId,
+      collaboratorTypeId,
+      linkBack,
+    }, {
+      ...(transaction && { transaction }),
+    });
+};
+
+/**
+ * Retrieves a entity collaborator record from the database.
+ *
+ * @param {string} genericCollaboratorType - entity type for collaborator.
+ * @param {object} sequelize - The Sequelize instance.
+ * @param {object} transaction - The transaction object.
+ * @param {number} entityId - The ID of the entity.
+ * @param {number} userId - The ID of the user.
+ * @param {string} typeName - The name of the collaborator type.
+ * @returns {Promise<object>} - A promise that resolves to the entity collaborator record.
+ */
+const getCollaboratorRecord = async (
+  genericCollaboratorType,
+  sequelize,
+  transaction,
+  entityId,
+  userId,
+  typeName,
+  // Find the entity collaborator record in the database
+) => sequelize.models[collaboratorDetails[genericCollaboratorType].collaborators]
+  .findOne({
+    where: {
+      [collaboratorDetails[genericCollaboratorType].idName]: entityId,
```

Comment:
Assuming I'm reading this correctly is there any concern for 'genericCollaboratorType' coming back undefined or empty then attempting to access a property like 'id'?

Reply:
i think it should be good